File: geomorphic.m
function newfield = geomorphic(field,par,dt)
%GEOMORPHIC operator accounting for bed erosion/deposition

% obtain norm and direction vectors:
vel = ( field.u.^2 + field.v.^2 ).^0.5;
ix = (vel>(par.g*par.h_min)^0.5).* field.u./max(vel,(par.g*par.h_min)^0.5);
iy = (vel>(par.g*par.h_min)^0.5).* field.v./max(vel,(par.g*par.h_min)^0.5);
% layer depths:
h = field.z_m - field.z_b;

% moving sediments
CH = h.*field.c_m;
%KH
KH = h.*field.k_m;

% SOLVE FOR CH
% start with explicit estimate
  Ze5 = ((par.alpha*field.k_m).^0.5/par.vs.*par.Rp^0.6.*((par.alpha*field.k_m).^0.5./par.g./max(h,par.h_min)).^0.08).^5;
  E = max( par.p*1.3e-7*par.vs*Ze5./(1+1.3e-7/0.3*Ze5) , 0 );
  D = max( par.vs*par.r0*field.c_m , 0 );
  CH_new = max(CH + dt*(E-D) , 0);
  h_new = max (h + dt/par.c_b*(E-D) , 0);
  C_new = max(CH_new./max(h_new,par.h_min) , 0);
  KH_new = KH - dt*0.5*par.R*par.g.*h_new.*(E-D);
  K_new = max(0 , KH_new./max(h_new,par.h_min));
% iterate 10 times with Newton scheme
if 1
for i = 1:10;
    Ze5 = ((par.alpha*K_new).^0.5/par.vs.*par.Rp^0.6.*((par.alpha*K_new).^0.5./par.g./max(h_new,par.h_min)).^0.08).^5;
    E = max( par.p*1.3e-7*par.vs*Ze5./(1+1.3e-7/0.3*Ze5) , 0 );
    D = max( par.vs*par.r0*C_new , 0 );
    CH_new = max(CH + dt*(E-D) , 0);
    h_new = max (h + dt/par.c_b*(E-D) , 0);
    C_new = max(CH_new./max(h_new,par.h_min) , 0);
    KH_new = KH - dt*0.5*par.R*par.g.*h_new.*(E-D);
    K_new = max(0 , KH_new./max(h_new,par.h_min));
end;
end;

% ensure dissipation of K (should we do that?)
% K_new = min(field.k_m, K_new);

% retrieve bed level change and impose limit. dzb is positive in case of deposition
dzb = min(dt.*(D-E)./par.c_b , (field.z_m-field.z_b).*field.c_m./par.c_b); % limited by deposition of all suspended sediments
dzb = max(dzb,field.z_r-field.z_b); % limit by rigid bottom

% retrieve all conservative variables from final bed level change
h_new = h-dzb;
CH_new = min(max(CH - dzb.*par.c_b , 0) , h_new) ;
KH_new = max(KH + 0.5*par.R*par.g.*h_new*par.c_b.*dzb , 0);

% Note 1: implication of erosion/deposition on momentum transfer to/from the
% bed is neglected.

% Note 2: Should we forbid a net gain in K in case of deposition?

% final update:
newfield = field;
newfield.z_b = field.z_b + dzb;
newfield.c_m = min ( max(CH_new./max(h_new,par.h_min) , 0) , 1);
newfield.c_m(h_new<par.h_min)=field.c_m(h_new<par.h_min); % do not change concentration where flow depth is about zero
newfield.k_m = max(KH_new./max(h_new,par.h_min) , 0);

File: geomorphic.py
# ENTRAINMENT operator accounting for water entrainment at the top of turbidity current
# solved implicitly with backward Euler scheme (solution obtained
# interatively from the explicit estimate with a Newton scheme)
import numpy as np
import sys

def geomorphic(field, par, dt):
    # (vel) calculates the norm of the velcocity vector at each point in the flow field, where (u) and (v) are the horizontal and vertical components
    vel = (field.u ** 2 + field.v ** 2) ** 0.5

    # direction vectors (ix) and (iy), repersent the horizontal and vertical components of the flow velocity 
    ix = ((vel > ((par.g * par.h_min) ** 0.5)).astype(int)) * (field.u / (np.maximum(vel, (par.g * par.h_min) ** 0.5)))
    iy = ((vel > ((par.g * par.h_min) ** 0.5)).astype(int)) * (field.v / (np.maximum(vel, (par.g * par.h_min) ** 0.5)))

    # (h) calculates current flow depth/thickness
    h = field.z_m - field.z_b

    # (CH) calculates the amount of sediment being transported by the current 
    CH = h * field.c_m

    # (KH) calculates the turbulent kinetic energy within the current
    KH = h * field.k_m

    Ze5 = ((par.alpha*field.k_m) ** 0.5 / par.vs * par.Rp ** 0.6 * ((par.alpha * field.k_m) ** 0.5 / par.g / np.maximum(h, par.h_min)) ** 0.08) ** 5
    E = np.maximum((par.p * (1.3 * 10 ** -7)) * par.vs * Ze5 / (1 + (1.3 * 10 ** -7) / 0.3 * Ze5), 0)
    D = np.maximum(par.vs * par.r0 * field.c_m, 0)
    CH_new = np.maximum(CH + dt * (E - D), 0)
    h_new = np.maximum(h + dt / par.c_b * (E - D), 0)
    C_new = np.maximum(CH_new / np.maximum(h_new, par.h_min), 0)
    KH_new = KH - dt * 0.5 * par.R * par.g * h_new * (E - D)
    K_new = np.maximum(0, KH_new / np.maximum(h_new, par.h_min))
    
    for i in range(10):
    
        Ze5 = ((par.alpha*K_new) ** 0.5 / par.vs * par.Rp ** 0.6 * ((par.alpha * K_new) ** 0.5 / par.g / np.maximum(h_new, par.h_min)) ** 0.08) ** 5
        E = np.maximum((par.p * (1.3 * 10 ** -7)) * par.vs * Ze5 / (1 + (1.3 * 10 ** -7) / 0.3 * Ze5), 0)
        D = np.maximum(par.vs * par.r0 * C_new, 0)
        CH_new = np.maximum(CH + dt * (E - D), 0)
        h_new = np.maximum(h + dt / par.c_b * ((E - D)), 0)
        C_new = np.maximum(CH_new / np.maximum(h_new, par.h_min), 0)
        KH_new = KH - dt * 0.5 * par.R * par.g * h_new * (E - D)
        K_new = np.maximum(0, KH_new / np.maximum(h_new, par.h_min))


    # retrieve bed level change and impose limit. dzb is positive in case of deposition
    dzb = np.minimum((dt * ((D - E)*7) / par.c_b), ((field.z_m - field.z_b) * field.c_m / par.c_b)) #limited by deposition of all suspended sediments
    dzb = np.maximum(dzb, field.z_r - field.z_b) # limit by rigid bottom

    # retrieve all conservative variables from final bed level change
    h_new = h - dzb
    CH_new = np.minimum(np.maximum(CH - dzb * par.c_b, 0), h_new)
    KH_new = np.maximum(KH + 0.5 * par.R * par.g * h_new * par.c_b * dzb, 0)


    # note 1: implication of erosion/deposition on momentum transfer to/form the bed is neglected
    # note 2: should we forbid a net gain in K in case of deposition
    # K factor is an empirical measure of soil erodibility as affected by intrinsic soil properties
    # https://www.sciencedirect.com/topics/earth-and-planetary-sciences/revised-universal-soil-loss-equation\
   
    # final update
    newfield = field
    newfield.z_b = field.z_b + dzb
    newfield.c_m = np.minimum(np.maximum(CH_new / np.maximum(h_new, par.h_min), 0), 1) # do not change concentration where flow depth is about zero
    if (np.any(np.isnan(h_new))):
        print('h_new after has nans')
        sys.exit()
    indices = np.where(h_new < par.h_min)
    newfield.c_m[indices] = field.c_m[indices]

    newfield.k_m = np.maximum(KH_new / np.maximum(h_new, par.h_min), 0)

    return newfield
